nexus/src/app/background/instance_watcher.rs

```diff
@@ -0,0 +1,635 @@
+            serde_json::json!({
+                "total_instances": total,
+                "instances_updated": instances_updated,
+                "vmms_updated": vmms_updated,
+                "no_change": no_change,
+                "not_found": not_found,
+                "unreachable_instances": unreachable_instances,
+                "sled_agent_errors": sled_agent_errors,
+                "check_errors": check_errors,
+                "pruned_instances": pruned,
+            })
+        }
+        .boxed()
+    }
+}
+
+mod metrics {
+    use super::{CheckError, CheckFailure, InstanceTarget, InstanceUpdated};
+    use oximeter::types::Cumulative;
+    use oximeter::Metric;
+    use oximeter::MetricsError;
+    use oximeter::Sample;
+    use std::collections::BTreeMap;
+    use std::sync::Arc;
+    use std::sync::Mutex;
+
+    #[derive(Debug, Default)]
+    pub(super) struct Metrics {
+        instances: BTreeMap<InstanceTarget, Instance>,
+    }
+
+    #[derive(Debug)]
+    pub(super) struct Producer(pub(super) Arc<Mutex<Metrics>>);
+
+    #[derive(Debug)]
+    pub(super) struct Instance {
+        no_update: InstanceChecks,
+        instance_updated: InstanceChecks,
+        vmm_updated: InstanceChecks,
+        both_updated: InstanceChecks,
+        check_failures: BTreeMap<CheckFailure, InstanceCheckFailures>,
+        update_failures: BTreeMap<CheckError, InstanceCheckErrors>,
+        touched: bool,
+    }
+
+    impl Metrics {
+        pub(crate) fn instance(
+            &mut self,
+            instance: InstanceTarget,
+        ) -> &mut Instance {
+            self.instances.entry(instance).or_insert_with(|| Instance {
+                no_update: InstanceChecks {
+                    instance_updated: false,
+                    vmm_updated: false,
+                    datum: Cumulative::default(),
+                },
+                instance_updated: InstanceChecks {
+                    instance_updated: true,
+                    vmm_updated: false,
+                    datum: Cumulative::default(),
+                },
+                vmm_updated: InstanceChecks {
+                    instance_updated: false,
+                    vmm_updated: true,
+                    datum: Cumulative::default(),
+                },
+                both_updated: InstanceChecks {
+                    instance_updated: true,
+                    vmm_updated: true,
+                    datum: Cumulative::default(),
+                },
+                check_failures: BTreeMap::new(),
+                update_failures: BTreeMap::new(),
+                touched: false,
+            })
+        }
+
+        pub(super) fn prune(&mut self) -> usize {
+            let mut pruned = 0;
+            self.instances.retain(|_, instance| {
+                let touched = std::mem::replace(&mut instance.touched, false);
+                if !touched {
+                    pruned += 1;
+                }
+                touched
+            });
```

Comment:
This is a bit of a style nit, but you could avoid the variable `pruned` here by computing the size at the start, then after the call to `retain()`, and then returning their difference.

Reply:
Yup, good catch — I think this was an artifact of a previous version of this code where I was `retain`ing a nested set of `BTreeMap`s and doing it that way was more complicated. I'll change it.